x-pack/plugin/security/qa/saml-rest-tests/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/saml/SamlServiceProviderMetadataIT.java

```diff
@@ -0,0 +1,254 @@
+                        .put(prefix + ".sp.acs", https + "acs/" + realmNumber)
+                        .put(prefix + ".attributes.principal", "urn:oid:2.5.4.3")
+                        .put(prefix + ".ssl.certificate_authorities", "ca.crt");
+
+                }
+                return settings.map();
+            })
+            .build();
+    }
+
+    private static void configureMetadataResource(int realmNumber) throws CertificateException, IOException, URISyntaxException {
+        metadataAvailable.putIfAbsent(realmNumber, false);
+
+        var signingCert = getDataResource("/saml/signing.crt");
+        var metadataBody = new SamlIdpMetadataBuilder().entityId(getIdpEntityId(realmNumber)).sign(signingCert).asString();
+        httpsServer.createContext("/metadata/" + realmNumber + ".xml", http -> {
+            if (metadataAvailable.get(realmNumber)) {
+                sendXmlContent(metadataBody, http);
+            } else {
+                if (randomBoolean()) {
+                    http.sendResponseHeaders(randomFrom(404, 401, 403, 500), 0);
+                } else {
+                    sendXmlContent("not valid xml", http);
+                }
+            }
+        });
+    }
+
+    private static void sendXmlContent(String bodyContent, HttpExchange http) throws IOException {
+        http.getResponseHeaders().add("Content-Type", "text/xml");
+        http.sendResponseHeaders(200, bodyContent.length());
+        try (var out = http.getResponseBody()) {
+            out.write(bodyContent.getBytes(StandardCharsets.UTF_8));
+        }
+    }
+
+    @BeforeClass
+    public static void loadCertificateAuthority() throws Exception {
+        URL resource = SamlServiceProviderMetadataIT.class.getResource("/ssl/ca.crt");
+        if (resource == null) {
+            throw new FileNotFoundException("Cannot find classpath resource /ssl/ca.crt");
+        }
+        caPath = PathUtils.get(resource.toURI());
+    }
+
+    @Override
+    protected String getTestRestCluster() {
+        return cluster.getHttpAddresses();
+    }
+
+    @Override
+    protected String getProtocol() {
+        return "https";
+    }
+
+    @Override
+    protected Settings restAdminSettings() {
+        final String token = basicAuthHeaderValue("test_admin", new SecureString("x-pack-test-password".toCharArray()));
+        return Settings.builder().put(ThreadContext.PREFIX + ".Authorization", token).put(CERTIFICATE_AUTHORITIES, caPath).build();
+    }
+
+    @Override
+    protected Settings restClientSettings() {
+        final String token = basicAuthHeaderValue("rest_test", new SecureString("rest_password".toCharArray()));
+        return Settings.builder().put(ThreadContext.PREFIX + ".Authorization", token).put(CERTIFICATE_AUTHORITIES, caPath).build();
+    }
+
+    public void testAuthenticationWhenMetadataIsUnreliable() throws Exception {
+        // Start with no metadata available
+        makeAllMetadataUnavailable();
```

Comment:
Is this necessary? The httpServer is configured to _not_ return metadata in the setup. If that is not the case, it will be too late here to mark them available because the ES cluster is already started. So I don't think this does much. Maybe we can replace it with an assertion that all metadata is indeed unavailable.